core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1167,19 +1122,9 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
      * @throws Exception On any error.
      */
     @throws[Exception]
-    def handleNewSession() {
-      info("ZK expired; shut down all controller components and try to re-elect")
-      if (controllerElector.getControllerID() != config.brokerId) {
-        onControllerResignation()
-        inLock(controllerContext.controllerLock) {
-          controllerElector.elect
-        }
-      } else {
-        // This can happen when there are multiple consecutive session expiration and handleNewSession() are called multiple
-        // times. The first call may already register the controller path using the newest ZK session. Therefore, the
-        // controller path will exist in subsequent calls to handleNewSession().
-        info("ZK expired, but the current controller id %d is the same as this broker id, skip re-elect".format(config.brokerId))
-      }
+    def handleNewSession(): Unit = {
+      controllerEventQueue.put(Resign(getControllerID()))
+      controllerEventQueue.put(Elect)
```

Comment:
I think it's possible but would prefer to not do it in this patch.

Here are the places where Resign and Elect are added to the queue:
1. KafkaController.startup adds Elect to the queue.
2. SessionExpirationListener.handleNewSession adds Resign and Elect to the queue.
3. ControllerChangeListener.handleDataChange adds Resign to the queue.
4. ControllerChangeListener.handleDataDeleted adds Resign and Elect to the queue.

I tried refactoring such that the above maintains the same behavior as today's KafkaController.startup, SessionExpirationListener.handleNewSession, LeaderChangeListener.handleDataChange, and LeaderChangeListener.handleDataDeleted. I'd like to try to maintain existing behavior in this patch wherever possible to minimize regressions.

If we merge the two into a single event, we'd end up doing some actions that would work but might not make sense:
1. KafkaController.startup would unnecessarily clear all state even though state is guaranteed to already be empty.
2. ControllerChangeListener.handleDataChange would run the election algorithm even though handleDataChange getting triggered should indicate that a new controller has already been elected without znode deletion (setData on /controller could have been used). It's not clear to me if there is some merit to doing this, as later controller changes would hopefully get picked up by ZkClient and enqueued.